File: app/server/routes/ai.ts
import type { Express, Request, Response } from "express";
import { z } from "zod";
import { storage } from "../storage";
import { isAuthenticated } from "../auth";
import { computeSuspects } from "../analytics/suspects";
import { synthesize } from "../ai/synthesis";
import { addDays, isIsoDate, localDate } from "../analytics/time";
import { resolveTimezone, sendZodError } from "./context";

const bodySchema = z.object({
  weekStart: z.string().optional(),
  symptom: z.string().max(80).optional().nullable(),
  tz: z.string().optional(),
});

export function registerAiRoutes(app: Express) {
  // AI Pattern Synthesis for the Food Suspect Digest
  app.post("/api/ai/synthesis", isAuthenticated, async (req: Request, res: Response) => {
    try {
      const body = bodySchema.parse(req.body ?? {});
      const tz = await resolveTimezone(req);
      const weekStart = body.weekStart || addDays(localDate(new Date(), tz), -6);
      if (!isIsoDate(weekStart)) return res.status(400).json({ message: "weekStart must be YYYY-MM-DD" });
      const symptom = body.symptom && body.symptom !== "All Symptoms" ? body.symptom : null;

      const userId = req.user!.id;
      const [meals, symptoms, correlations, settings, watchlist] = await Promise.all([
        storage.getMealsByUser(userId),
        storage.getSymptomsByUser(userId),
        storage.getAllCorrelationsByUser(userId),
        storage.getSettings(userId),
        storage.getWatchlist(userId),
      ]);
      const suspects = computeSuspects(meals, symptoms, correlations, watchlist.map((w) => w.ingredient), weekStart, tz, settings.correlationWindowHours, symptom);
      res.json(await synthesize(userId, suspects, symptom));
    } catch (error) {
      if (error instanceof z.ZodError) {
        sendZodError(res, "Invalid request", error);
      } else {
        console.error("Error generating synthesis:", error);
        res.status(500).json({ message: "Failed to generate synthesis" });
      }
    }
  });
}

File: api/app/routers/ai.py
"""The AI summary on the Food Suspect Digest."""

from datetime import UTC, datetime

from fastapi import APIRouter, HTTPException, status

from app.ai.synthesis import synthesize
from app.deps import CurrentUser, Session, get_settings_row, resolve_timezone
from app.domain.suspects import ALL_SYMPTOMS, compute_suspects
from app.domain.time import add_days, is_iso_day, local_date
from app.schemas import SynthesisRead, SynthesisRequest
from app.services.rows import correlation_row, load_correlations, load_history, settings_row
from app.services.watchlist import watchlist_ingredients

router = APIRouter(tags=["ai"])


@router.post("/ai/synthesis", response_model=SynthesisRead)
async def synthesis(body: SynthesisRequest, user: CurrentUser, session: Session) -> dict:
    zone = await resolve_timezone(session, user, body.tz)
    week_start = body.week_start or add_days(local_date(datetime.now(UTC), zone), -6)
    if not is_iso_day(week_start):
        raise HTTPException(status.HTTP_400_BAD_REQUEST, detail="weekStart must be YYYY-MM-DD")

    chosen = body.symptom if body.symptom and body.symptom != ALL_SYMPTOMS else None
    settings = await get_settings_row(session, user)
    meals, symptoms = await load_history(session, user.id)
    correlations = [correlation_row(c) for c in await load_correlations(session, user.id)]
    watchlist = await watchlist_ingredients(session, user.id)

    suspects = compute_suspects(
        meals, symptoms, correlations, watchlist, week_start, zone,
        settings.correlation_window_hours, chosen,
    )
    return await synthesize(session, user.id, suspects, chosen)
